web/book/src/internals/modules.md

```diff
@@ -0,0 +1,311 @@
+# Modules
+
+> Status: proposal
+
+> This is a technical document. For a "how to use" or a TLDR; skip to the
+> [Example](#Example) section.
+
+Design goals:
+
+1. Allow importing declarations from other files.
+
+2. Have namespaces for things like `std`.
+
+3. Have a hierarchical structure so we can represent files in directories.
+
+4. Have unambiguous module structure within a project.
+
+## Definition
+
+A module is a namespace that contains declarations. A module is itself a
+declaration , which means that it can contain nested child modules.
+
+This means that modules form a
+[tree graph](<https://en.wikipedia.org/wiki/Tree_(graph_theory)>), which we call
+"the module structure".
+
+For the sake of this document, we will express the module structure with
+`module` keyword and a code block encased in curly braces:
+
+```
+module my_playlists {
+    let bangers = ... # a declaration
+
+    module soundtracks {
+        let movie_albums = ... # another declaration
+    }
+}
+```
+
+> The syntax `module name { ...decls... }` is not part of PRQL language, with
+> the objection that it is unnecessary as it only adds more ways of defining
+> modules. If a significant upside of this syntax is found, it may be added in
+> the future.
+
+## Name resolution
+
+Any declarations within a module can be referenced from the outside of the
+module:
+
+```prql no-eval
+# using module structure declared above
+
+let great_tracks = my_playlists.bangers
+
+let movie_scores = my_playlists.soundtracks.movie_albums
+```
+
+Identifiers are resolved relative to current module.
+
+```prql no-eval
+module my_playlists {
+    module soundtracks {
+        let movie_albums = (from albums | filter id == 3)
+    }
+
+    from soundtracks.movie_albums
+}
+from my_playlists.soundtracks.movie_albums
+```
+
+If an identifier cannot be resolved relative to the current module, it tries to
+resolve relative to the parent module. This is repeated, stepping up the module
+hierarchy until a match is found or root of the module structure is reached.
+
+```prql no-eval
+module my_playlists {
+    let decl_1 = ...
+
+    module soundtracks {
+        let decl_2 = ...
+    }
+
+    module upbeat_rock {
+        let decl_3 = ...
+
+        from decl_1 | join soundtracks.decl2 | join decl_3
+    }
+}
+```
+
+## Main var declaration
+
+If last declaration in a module is a variable declaration that is named `main`,
+then the leading `let main = ` can be omitted and expressed only by the
+expression itself.
+
+```
+module my_playlists {
+    let bangers = (from tracks | take 10)
+
+    from playlists | take 10
+}
+
+let album_titles = my_playlists.main
+```
+
+When a module is referenced as a value, the `main` variable is used instead.
+This is especially useful when referring to a module which is to be compiled to
+RQ (and later SQL).
+
+## File importing
```

Comment:
What's the advantage of this relative to just including all PRQL files as modules automatically, like python does?

Reply:
What do you mean include automatically? In Python you still have to have an "import x from y" for each file. 

Even more so, if file a.py imports b.py file, some other file c.py has to re-import b.py. With our spec, c.py could just refer to `a.b` to get to the contents of b.py.

Reply:
OK, that makes sense — but in that case I'm confused by something else; I'll move the comment there